### PR TITLE
Support fetching entities by aliased name

### PR DESCRIPTION
**Pull request by Ocramius**
See #385 (this PR supersedes it)

The final aim is allowing something like:

``` php
$user = $em->find(UserInterface::class, 123);
```

This PR depends also on doctrine/common#342

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -18,6 +18,7 @@
         "doctrine/collections": "~1.2",
         "doctrine/dbal": ">=2.5-dev,<2.6-dev",
         "doctrine/instantiator": "~1.0.1",
+        "doctrine/common": ">=2.5-dev,<2.6-dev",
```
Comment (stof) [inline composer.json]: indeed, it must be there

File: lib/Doctrine/ORM/Event/OnClassMetadataNotFoundEventArgs.php [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+use Doctrine\ORM\EntityManager;
+use Doctrine\ORM\EntityManagerInterface;
+
+/**
+ * Class that holds event arguments for a loadMetadata event.
+ *
+ * @author Jonathan H. Wage <jonwage@gmail.com>
+ * @since  2.3
+ */
+class OnClassMetadataNotFoundEventArgs extends EventArgs
```
Comment (stof) [inline lib/Doctrine/ORM/Event/OnClassMetadataNotFoundEventArgs.php]: why not extending ``Doctrine\Common\Persistence\Event\ManagerEventArgs`` ?

File: lib/Doctrine/ORM/Event/OnClassMetadataNotFoundEventArgs.php [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    /**
+     * @return ClassMetadata|null
+     */
+    public function getFoundMetadata()
+    {
+        return $this->foundMetadata;
+    }
+
+    /**
+     * Retrieve associated ClassMetadata.
```
Comment (deeky666) [inline lib/Doctrine/ORM/Event/OnClassMetadataNotFoundEventArgs.php]: This description is wrong, isn't it?

File: lib/Doctrine/ORM/Events.php [hunk 4]
```diff
@@ -121,6 +121,14 @@ private function __construct()
     const loadClassMetadata = 'loadClassMetadata';
 
     /**
+     * The preLoadClassMetadata event occurs before the mapping metadata for a
```
Comment (deeky666) [inline lib/Doctrine/ORM/Events.php]: The description here is wrong, too. Copy-paste?

File: lib/Doctrine/ORM/Mapping/ClassMetadataFactory.php [hunk 5]
```diff
@@ -91,6 +92,22 @@ protected function initialize()
     /**
      * {@inheritDoc}
```
Comment (deeky666) [inline lib/Doctrine/ORM/Mapping/ClassMetadataFactory.php]: There is no parent method here as far as I can see. Can you provide a descriptive DocBlock?

Comment (Ocramius) [inline lib/Doctrine/ORM/Mapping/ClassMetadataFactory.php]: The method is actually an override as per doctrine/common#342

**Pull request merged**